# Data Wrangling

## Dataset - WeRateDogs&trade; Twitter Archive

***By: Kartik Nanduri***<br>
**Date: 1st Dec, 2018.**

## Let's Gather

In [1]:
# importing all the necessary libraries
import os
import pandas as pd
import requests as req

***<span style="color: red">Important! uncomment the following files to run the book with out errors.</span>***

In [2]:
# resetting the folder structure.
#os.rename('dataset/twitter-archive-enhanced.csv', 'twitter-archive-enhanced.csv')
#import shutil
#shutil.rmtree('dataset')

***<span style="color: green">Important! once done, please recomment.</span>***

1. [x] **The file given at hand `twitter-archive-enhanced.csv`**

In [3]:
# all the requried files for this project are in the list files_list
files_list = ['twitter-archive-enhanced.csv', 'image-predictions.tsv', 'tweet_json.txt']

In [4]:
# reading the twitter archive file
archive = pd.read_csv(files_list[0])

# taking at random entries for the archive file
archive.sample(2)

tweet_id  in_reply_to_status_id  in_reply_to_user_id  \
1512  691321916024623104                    NaN                  NaN   
534   807106840509214720                    NaN                  NaN   

                      timestamp  \
1512  2016-01-24 18:09:23 +0000   
534   2016-12-09 06:17:20 +0000   

                                                 source  \
1512  <a href="http://twitter.com/download/iphone" r...   
534   <a href="http://twitter.com/download/iphone" r...   

                                                   text  retweeted_status_id  \
1512  This is Molly. She's a Peruvian Niddlewog. Lov...                  NaN   
534   This is Stephan. He just wants to help. 13/10 ...                  NaN   

      retweeted_status_user_id retweeted_status_timestamp  \
1512                       NaN                        NaN   
534                        NaN                        NaN   

                                          expanded_urls  rating_numerator  \
1512  https://twitter.com/dog_rates/status/691321916...                11   
534   https://twitter.com/dog_rates/status/807106840...                13   

      rating_denominator     name doggo floofer pupper puppo  
1512                  10    Molly  None    None   None  None  
534                   10  Stephan  None    None   None  None

2. [x] **Fetching the data from url and saving it to local drive - `image-predictions.tsv`**

In [5]:
# reading the file from internet using the requests library
url = "https://s3.amazonaws.com/video.udacity-data.com/topher/2018/November/5bf60c69_image-predictions-3/image-predictions-3.tsv"
res = req.get(url)

with open(files_list[1], mode = "wb") as op_file:
    op_file.write(res.content)

In [6]:
# checking if fetched the data right way
img_pre_test = pd.read_csv(files_list[1], delimiter = "\t", encoding = 'utf-8')
img_pre_test.sample(2)

# we did it the right way, Yay! it worked.

tweet_id                                          jpg_url  \
104   667832474953625600  https://pbs.twimg.com/media/CUSeGFNW4AAyyHC.jpg   
1603  800141422401830912  https://pbs.twimg.com/media/CxqsX-8XUAAEvjD.jpg   

      img_num                  p1   p1_conf  p1_dog          p2   p2_conf  \
104         1  miniature_pinscher  0.214200    True  bath_towel  0.146789   
1603        3    golden_retriever  0.938048    True      kuvasz  0.025119   

      p2_dog                  p3   p3_conf  p3_dog  
104    False           Chihuahua  0.104152    True  
1603    True  Labrador_retriever  0.022977    True

3. [x] **Getting data from Twitter&trade;**

In [7]:
# importing all the necessary libraries for accessing Twitter via API
import tweepy
from tweepy import OAuthHandler
import json
from timeit import default_timer as timer

In [8]:
# setting up all the necessary placeholders for API
consumer_key = 'xxx.xxx.xxx.xxx'
consumer_secret = 'xxx.xxx.xxx.xxx'
access_token = 'xxx.xxx.xxx.xxx'
access_secret = 'xxx.xxx.xxx.xxx'

auth = OAuthHandler(consumer_key, consumer_secret)
auth.set_access_token(access_token, access_secret)

api = tweepy.API(auth_handler = auth,
                 parser = tweepy.parsers.JSONParser(),
                 wait_on_rate_limit = True, 
                 wait_on_rate_limit_notify = True)

In [9]:
def fetch_and_save(ids, api_ins, one_id = None):
    '''
    This function will fetch data with associated id in ids list
    ids (List Object): a list all tweets
    api_ins (Tweepy Object): api object instance, will be used to query twitter for data
    one_id (int): use when you want to query only for one tweet
    failed_ids (List Object): a list will be retured so that, this fuction can be called once again on those ids
    '''
    new_file_name = ''; failed_ids = []; tweet_df = []
    
    # checking if file exists
    if os.path.exists(files_list[2]):
        temp = [s for s in os.listdir() if "tweet_json" in s]
        new_file_name = files_list[2].split('.')[0] + "_" + str(len(temp)) + ".txt"
    else:
        new_file_name = files_list[2]
    
    # querying a list of ids
    if one_id == None:
        with open(new_file_name, mode = 'w') as outfile:
            for one_id in ids:
                try:
                    content = api_ins.get_status(one_id, tweet_mode='extended')
                    json.dump(content, outfile)
                    outfile.write('\n')
                
                except Exception as e:
                    print("Error for: " + str(one_id) + " - " + str(e))
                    failed_ids.append(one_id)
    
    # querying a single id
    else:
        try:
            content = api_ins.get_status(one_id, include_entities = True)
            favorites = content['favorite_count']
            retweets = content['retweet_count']
            
            tweet_df.append({'tweet_id': int(one_id),
                        'favorites': int(favorites),
                        'retweets': int(retweets)})
            
            return tweet_df
                           
        except Exception as e:
            print("Error for: " + str(one_id) + " - " + str(e))
            failed_ids.append(one_id)

    return failed_ids

In [10]:
# passing the list of ids to the fuction fetch_and_save()
tweet_ids = archive['tweet_id'].tolist()
len(tweet_ids)

2356

In [11]:
# fetching data
# starting the timer
start = timer()

# querying
errors = fetch_and_save(tweet_ids, api)

# ending the timer
end = timer()

# calculating the runtime for fetch_and_save
print("That took about {} mins.".format(round((end - start)/60, 1)))

Error for: 888202515573088257 - [{'code': 144, 'message': 'No status found with that ID.'}]
Error for: 873697596434513921 - [{'code': 144, 'message': 'No status found with that ID.'}]
Error for: 872668790621863937 - [{'code': 144, 'message': 'No status found with that ID.'}]
Error for: 869988702071779329 - [{'code': 144, 'message': 'No status found with that ID.'}]
Error for: 866816280283807744 - [{'code': 144, 'message': 'No status found with that ID.'}]
Error for: 861769973181624320 - [{'code': 144, 'message': 'No status found with that ID.'}]
Error for: 845459076796616705 - [{'code': 144, 'message': 'No status found with that ID.'}]
Error for: 842892208864923648 - [{'code': 144, 'message': 'No status found with that ID.'}]
Error for: 837012587749474308 - [{'code': 144, 'message': 'No status found with that ID.'}]
Error for: 827228250799742977 - [{'code': 144, 'message': 'No status found with that ID.'}]
Error for: 802247111496568832 - [{'code': 144, 'message': 'No status found with 

Rate limit reached. Sleeping for: 502


Error for: 758740312047005698 - Failed to send request: ('Connection aborted.', OSError("(10054, 'WSAECONNRESET')"))
Error for: 754011816964026368 - [{'code': 144, 'message': 'No status found with that ID.'}]


Rate limit reached. Sleeping for: 515


Error for: 676957860086095872 - Failed to send request: ('Connection aborted.', OSError("(10054, 'WSAECONNRESET')"))
That took about 34.8 mins.


In [12]:
# lets save the failed ids into one master list
print("Total failed request are: {}. \n".format(len(errors)))

# ids that failed and the ones that passed
indi_fail = []; success = []

#for each failed id, lets try to fetch status individually.
for error in errors:
    temp = fetch_and_save(ids = None, api_ins = api, one_id = error)
    indi_fail.append(temp[0])

# removing empty elements from list
success = [x for x in indi_fail if not isinstance(x, (int))]
indi_fail = [x for x in indi_fail if isinstance(x, (int))]

# checking if there is change
print("\nWe were able to retrieve {} records, others failed.".format(len(errors) - len(indi_fail)))

Total failed request are: 16. 

Error for: 888202515573088257 - [{'code': 144, 'message': 'No status found with that ID.'}]
Error for: 873697596434513921 - [{'code': 144, 'message': 'No status found with that ID.'}]
Error for: 872668790621863937 - [{'code': 144, 'message': 'No status found with that ID.'}]
Error for: 869988702071779329 - [{'code': 144, 'message': 'No status found with that ID.'}]
Error for: 866816280283807744 - [{'code': 144, 'message': 'No status found with that ID.'}]
Error for: 861769973181624320 - [{'code': 144, 'message': 'No status found with that ID.'}]
Error for: 845459076796616705 - [{'code': 144, 'message': 'No status found with that ID.'}]
Error for: 842892208864923648 - [{'code': 144, 'message': 'No status found with that ID.'}]
Error for: 837012587749474308 - [{'code': 144, 'message': 'No status found with that ID.'}]
Error for: 827228250799742977 - [{'code': 144, 'message': 'No status found with that ID.'}]
Error for: 802247111496568832 - [{'code': 144, '

In [13]:
# printing the results of success
success

[{'tweet_id': 758740312047005698, 'favorites': 6076, 'retweets': 1741},
 {'tweet_id': 676957860086095872, 'favorites': 2302, 'retweets': 857}]

4. [x] **Okay, let's read `retweets` and `favourite count` from `tweet_json.txt`**

In [14]:
# reading tweet_json.txt
tweets = pd.read_json(files_list[2], lines = True, encoding = 'utf-8')

In [15]:
# let's select only the following columns retweet_count, favorite_count, id
tweets = tweets[['id', 'favorite_count', 'retweet_count']]
tweets.info()

<class 'pandas.core.frame.DataFrame'>
RangeIndex: 2340 entries, 0 to 2339
Data columns (total 3 columns):
id                2340 non-null int64
favorite_count    2340 non-null int64
retweet_count     2340 non-null int64
dtypes: int64(3)
memory usage: 54.9 KB


In [16]:
# renaming the columns
tweets.rename(columns={'id': 'tweet_id',
                       'favorite_count': 'favorites',
                       'retweet_count': 'retweets'},
              inplace = True)

In [17]:
# concating the dataframes into one
success = pd.DataFrame(success, columns = ['tweet_id',
                                           'favorites',
                                           'retweets'])
tweet_master = pd.concat([tweets, success],
                         ignore_index = True,
                         sort = True,)
tweet_master.info()

<class 'pandas.core.frame.DataFrame'>
RangeIndex: 2342 entries, 0 to 2341
Data columns (total 3 columns):
favorites    2342 non-null int64
retweets     2342 non-null int64
tweet_id     2342 non-null int64
dtypes: int64(3)
memory usage: 55.0 KB


In [18]:
# making a copy of the archive
archive_copy = archive.copy()

# marking the id null that we failed to retrieve in archive
for a_id in indi_fail:
    archive_copy.loc[archive_copy['tweet_id'] == a_id, ['tweet_id']] = 0

In [19]:
# a checking if we did it right
len(archive_copy[archive_copy['tweet_id'] == 0])

14

In [20]:
# appending the new file to our files list
files_list.append('archive_copy.csv')

In [21]:
# writing the contents of tweet_master to a file
import numpy as np
tweet_master['tweet_id'] = tweet_master['tweet_id'].astype(np.int64)
tweet_master.to_csv('tweet_master.csv', index = False)

# saving the updated version of our archived-enhanced.csv
archive_copy['tweet_id'] = archive_copy['tweet_id'].astype(np.int64)
archive_copy.to_csv(files_list[3], index = False)

5. [x] **Last thing to do is to tidy up our folder, let's get going.**

In [22]:
# moving all data files under one folder - dataset
# removing the temporary files, that acted as placeholders

# creating the folder
folder = 'dataset'
if not os.path.exists(folder):
    os.mkdir(folder)

# we know that our master datasets for this project are
# 1. twitter-archive-enhanced.csv
# 2. image-predictions.tsv
# 3. tweet_json.txt
# 4. tweet_master.csv
# let us move these files

# updating our files_list
files_list.append('tweet_master.csv')

# moving only required files
for file in files_list:
    if os.path.exists(file):
        os.rename(file, folder+'/'+file)
    
# lisitng the current directory
os.listdir()

# clean and neat, lets get with assessing and cleaning

['.git',
 '.ipynb_checkpoints',
 'act_report.ipynb',
 'dataset',
 'error.png',
 'New Text Document.txt',
 'README.md',
 'wrangle_act.ipynb']

In [23]:
# renaming files_list
for i in range(len(files_list)):
    files_list[i] = folder + '/'+ files_list[i]

## Summary - Gathering

- We know, that gathering is a the first step in wrangling.
- We were successful in gathering from three different sources with different techniques:
    - Data given at hand.
    - Fetch from flat file stored on a server.
    - From API.

- There a total of 14 missing data points, tried a different ways for retrieving them, using the API as well as `twurl` of the `Ruby` package, but they were not to be found, as stated below in the highlighted section.

***<span style="color: ##6c6cff">So let's start with assessing the data.</span>***

![error](error.png)

## Assessing

In [24]:
files_list

['dataset/twitter-archive-enhanced.csv',
 'dataset/image-predictions.tsv',
 'dataset/tweet_json.txt',
 'dataset/archive_copy.csv',
 'dataset/tweet_master.csv']

In [25]:
# let's load up dataset, and starting assessing them.
archive =  pd.read_csv(files_list[-2], encoding = 'utf-8')
img_pre = pd.read_csv(files_list[1], sep = '\t', encoding = 'utf-8')
tweet_master = pd.read_csv(files_list[-1], encoding =  'utf-8')

### Issues to sort!

In [26]:
# printing out archive - visual assessment
archive

tweet_id  in_reply_to_status_id  in_reply_to_user_id  \
0     892420643555336193                    NaN                  NaN   
1     892177421306343426                    NaN                  NaN   
2     891815181378084864                    NaN                  NaN   
3     891689557279858688                    NaN                  NaN   
4     891327558926688256                    NaN                  NaN   
5     891087950875897856                    NaN                  NaN   
6     890971913173991426                    NaN                  NaN   
7     890729181411237888                    NaN                  NaN   
8     890609185150312448                    NaN                  NaN   
9     890240255349198849                    NaN                  NaN   
10    890006608113172480                    NaN                  NaN   
11    889880896479866881                    NaN                  NaN   
12    889665388333682689                    NaN                  NaN   
13    889638837579907072                    NaN                  NaN   
14    889531135344209921                    NaN                  NaN   
15    889278841981685760                    NaN                  NaN   
16    888917238123831296                    NaN                  NaN   
17    888804989199671297                    NaN                  NaN   
18    888554962724278272                    NaN                  NaN   
19                     0                    NaN                  NaN   
20    888078434458587136                    NaN                  NaN   
21    887705289381826560                    NaN                  NaN   
22    887517139158093824                    NaN                  NaN   
23    887473957103951883                    NaN                  NaN   
24    887343217045368832                    NaN                  NaN   
25    887101392804085760                    NaN                  NaN   
26    886983233522544640                    NaN                  NaN   
27    886736880519319552                    NaN                  NaN   
28    886680336477933568                    NaN                  NaN   
29    886366144734445568                    NaN                  NaN   
...                  ...                    ...                  ...   
2326  666411507551481857                    NaN                  NaN   
2327  666407126856765440                    NaN                  NaN   
2328  666396247373291520                    NaN                  NaN   
2329  666373753744588802                    NaN                  NaN   
2330  666362758909284353                    NaN                  NaN   
2331  666353288456101888                    NaN                  NaN   
2332  666345417576210432                    NaN                  NaN   
2333  666337882303524864                    NaN                  NaN   
2334  666293911632134144                    NaN                  NaN   
2335  666287406224695296                    NaN                  NaN   
2336  666273097616637952                    NaN                  NaN   
2337  666268910803644416                    NaN                  NaN   
2338  666104133288665088                    NaN                  NaN   
2339  666102155909144576                    NaN                  NaN   
2340  666099513787052032                    NaN                  NaN   
2341  666094000022159362                    NaN                  NaN   
2342  666082916733198337                    NaN                  NaN   
2343  666073100786774016                    NaN                  NaN   
2344  666071193221509120                    NaN                  NaN   
2345  666063827256086533                    NaN                  NaN   
2346  666058600524156928                    NaN                  NaN   
2347  666057090499244032                    NaN                  NaN   
2348  666055525042405380                    NaN                  NaN   
2349  666051853826850816                    NaN         

In [27]:
# Programmatic Assessment 1 - Information
archive.info()

<class 'pandas.core.frame.DataFrame'>
RangeIndex: 2356 entries, 0 to 2355
Data columns (total 17 columns):
tweet_id                      2356 non-null int64
in_reply_to_status_id         78 non-null float64
in_reply_to_user_id           78 non-null float64
timestamp                     2356 non-null object
source                        2356 non-null object
text                          2356 non-null object
retweeted_status_id           181 non-null float64
retweeted_status_user_id      181 non-null float64
retweeted_status_timestamp    181 non-null object
expanded_urls                 2297 non-null object
rating_numerator              2356 non-null int64
rating_denominator            2356 non-null int64
name                          2356 non-null object
doggo                         2356 non-null object
floofer                       2356 non-null object
pupper                        2356 non-null object
puppo                         2356 non-null object
dtypes: float64(4), int64(3), ob

In [28]:
# Programmtic Assessment 2 - Describe
archive.describe()

tweet_id  in_reply_to_status_id  in_reply_to_user_id  \
count  2.356000e+03           7.800000e+01         7.800000e+01   
mean   7.378107e+17           7.455079e+17         2.014171e+16   
std    8.885894e+16           7.582492e+16         1.252797e+17   
min    0.000000e+00           6.658147e+17         1.185634e+07   
25%    6.776996e+17           6.757419e+17         3.086374e+08   
50%    7.178159e+17           7.038708e+17         4.196984e+09   
75%    7.986755e+17           8.257804e+17         4.196984e+09   
max    8.924206e+17           8.862664e+17         8.405479e+17   

       retweeted_status_id  retweeted_status_user_id  rating_numerator  \
count         1.810000e+02              1.810000e+02       2356.000000   
mean          7.720400e+17              1.241698e+16         13.126486   
std           6.236928e+16              9.599254e+16         45.876648   
min           6.661041e+17              7.832140e+05          0.000000   
25%           7.186315e+17              4.196984e+09         10.000000   
50%           7.804657e+17              4.196984e+09         11.000000   
75%           8.203146e+17              4.196984e+09         12.000000   
max           8.874740e+17              7.874618e+17       1776.000000   

       rating_denominator  
count         2356.000000  
mean            10.455433  
std              6.745237  
min              0.000000  
25%             10.000000  
50%             10.000000  
75%             10.000000  
max            170.000000

In [29]:
# checking for duplicates - tweet_ids
# these are the 0's for which the api failed to retrive data
sum(archive.tweet_id.duplicated())

13

In [30]:
# checkin if we have more than one class of dogs assigned to dog
# the following are the only combinations that are present in the dataset
cond_1 = (archive['doggo'] == 'doggo') & (archive['floofer'] == 'floofer')
cond_2 = (archive['doggo'] == 'doggo') & (archive['pupper'] == 'pupper')
cond_3 = (archive['doggo'] == 'doggo') & (archive['puppo'] == 'puppo')

# printing these entries
archive[cond_1 | cond_2 | cond_3][['tweet_id', 'text', 'doggo', 'floofer', 'pupper', 'puppo']]

tweet_id                                               text  \
191   855851453814013952  Here's a puppo participating in the #ScienceMa...   
200   854010172552949760  At first I thought this was a shy doggo, but i...   
460   817777686764523521  This is Dido. She's playing the lead role in "...   
531   808106460588765185  Here we have Burke (pupper) and Dexter (doggo)...   
565   802265048156610565  Like doggo, like pupper version 2. Both 11/10 ...   
575   801115127852503040  This is Bones. He's being haunted by another d...   
705   785639753186217984  This is Pinot. He's a sophisticated doggo. You...   
733   781308096455073793  Pupper butt 1, Doggo 0. Both 12/10 https://t.c...   
778   775898661951791106  RT @dog_rates: Like father (doggo), like son (...   
822   770093767776997377  RT @dog_rates: This is just downright precious...   
889   759793422261743616  Meet Maggie &amp; Lila. Maggie is the doggo, L...   
956   751583847268179968  Please stop sending it pictures that don't eve...   
1063  741067306818797568  This is just downright precious af. 12/10 for ...   
1113  733109485275860992  Like father (doggo), like son (pupper). Both 1...   

      doggo  floofer  pupper  puppo  
191   doggo     None    None  puppo  
200   doggo  floofer    None   None  
460   doggo     None  pupper   None  
531   doggo     None  pupper   None  
565   doggo     None  pupper   None  
575   doggo     None  pupper   None  
705   doggo     None  pupper   None  
733   doggo     None  pupper   None  
778   doggo     None  pupper   None  
822   doggo     None  pupper   None  
889   doggo     None  pupper   None  
956   doggo     None  pupper   None  
1063  doggo     None  pupper   None  
1113  doggo     None  pupper   None

In [31]:
len(archive[cond_1 | cond_2 | cond_3])

14

1. **`twitter-archive-enhanced.csv`** table

***1 Content Issues:***

**1.1 Visual Assessment:**
- `rating_numerator` : has values such as 1, 3.. e.t.c - **Data Quality Dimension - `Consistency`**.
- `rating_denominator` : have values, less than 10, for example, the tweet_id - 666287406224695296 has the number 2 as its value - **Data Quality Dimension - `Consistency`**. 
- We see that, Articles - `a`, `an`, `the` have been used to name dogs, as well as words such as `such`, `quite` - **Data Quality Dimension - `Validity`**.
- There are instances where the names of dogs are in lowercases - **Data Quality Dimension - `Consistency`**.

**1.2 Programmatic Assessment:**
- `rating_numerator` : has a maximum value of 1766 - **Data Quality Dimension - `Consistency`**. 
- `rating_denominator` : has a maximum value of 170 - **Data Quality Dimension - `Consistency`**.
- All in all, this dataset appears to be clean, except for `expanded_url` - we have about 59 instances missing - **Data Quality Dimension - `Completeness`**.
- We can see that there are more than one class assigned to tweets, analyze and assign proper dog class so that melting is easy - **Data Quality Dimension - `Consistency`**.

***2 Structural Issues:***

**2.1 Visual Assessment:**
- we can see that, there are four classes of dogs `doggo`, `floofer`, `puppo`, `pupper`; these should a part of one unit - `dog_class` - **Data Quality Dimension - `Consistency`**.

**2.2 Programmatic Assessment:**
- `in_reply_to_status_id`, `retweeted_status_id`, `retweeted_status_user_id`, `in_reply_to_user_id` of type float64 must be converted into int - **Data Quality Dimension - `Validity`**.
- `timestamp`, `retweeted_status_timestamp` of type object must be converted into datatime - **Data Quality Dimension - `Validity`**.

In [32]:
# assessing img_predictions dataset
img_pre

tweet_id                                            jpg_url  \
0     666020888022790149    https://pbs.twimg.com/media/CT4udn0WwAA0aMy.jpg   
1     666029285002620928    https://pbs.twimg.com/media/CT42GRgUYAA5iDo.jpg   
2     666033412701032449    https://pbs.twimg.com/media/CT4521TWwAEvMyu.jpg   
3     666044226329800704    https://pbs.twimg.com/media/CT5Dr8HUEAA-lEu.jpg   
4     666049248165822465    https://pbs.twimg.com/media/CT5IQmsXIAAKY4A.jpg   
5     666050758794694657    https://pbs.twimg.com/media/CT5Jof1WUAEuVxN.jpg   
6     666051853826850816    https://pbs.twimg.com/media/CT5KoJ1WoAAJash.jpg   
7     666055525042405380    https://pbs.twimg.com/media/CT5N9tpXIAAifs1.jpg   
8     666057090499244032    https://pbs.twimg.com/media/CT5PY90WoAAQGLo.jpg   
9     666058600524156928    https://pbs.twimg.com/media/CT5Qw94XAAA_2dP.jpg   
10    666063827256086533    https://pbs.twimg.com/media/CT5Vg_wXIAAXfnj.jpg   
11    666071193221509120    https://pbs.twimg.com/media/CT5cN_3WEAAlOoZ.jpg   
12    666073100786774016    https://pbs.twimg.com/media/CT5d9DZXAAALcwe.jpg   
13    666082916733198337    https://pbs.twimg.com/media/CT5m4VGWEAAtKc8.jpg   
14    666094000022159362    https://pbs.twimg.com/media/CT5w9gUW4AAsBNN.jpg   
15    666099513787052032    https://pbs.twimg.com/media/CT51-JJUEAA6hV8.jpg   
16    666102155909144576    https://pbs.twimg.com/media/CT54YGiWUAEZnoK.jpg   
17    666104133288665088    https://pbs.twimg.com/media/CT56LSZWoAAlJj2.jpg   
18    666268910803644416    https://pbs.twimg.com/media/CT8QCd1WEAADXws.jpg   
19    666273097616637952    https://pbs.twimg.com/media/CT8T1mtUwAA3aqm.jpg   
20    666287406224695296    https://pbs.twimg.com/media/CT8g3BpUEAAuFjg.jpg   
21    666293911632134144    https://pbs.twimg.com/media/CT8mx7KW4AEQu8N.jpg   
22    666337882303524864    https://pbs.twimg.com/media/CT9OwFIWEAMuRje.jpg   
23    666345417576210432    https://pbs.twimg.com/media/CT9Vn7PWoAA_ZCM.jpg   
24    666353288456101888    https://pbs.twimg.com/media/CT9cx0tUEAAhNN_.jpg   
25    666362758909284353    https://pbs.twimg.com/media/CT9lXGsUcAAyUFt.jpg   
26    666373753744588802    https://pbs.twimg.com/media/CT9vZEYWUAAlZ05.jpg   
27    666396247373291520    https://pbs.twimg.com/media/CT-D2ZHWIAA3gK1.jpg   
28    666407126856765440    https://pbs.twimg.com/media/CT-NvwmW4AAugGZ.jpg   
29    666411507551481857    https://pbs.twimg.com/media/CT-RugiWIAELEaq.jpg   
...                  ...                                                ...   
2045  886366144734445568    https://pbs.twimg.com/media/DE0BTnQUwAApKEH.jpg   
2046  886680336477933568    https://pbs.twimg.com/media/DE4fEDzWAAAyHMM.jpg   
2047  886736880519319552    https://pbs.twimg.com/media/DE5Se8FXcAAJFx4.jpg   
2048  886983233522544640    https://pbs.twimg.com/media/DE8yicJW0AAAvBJ.jpg   
2049  887101392804085760    https://pbs.twimg.com/media/DE-eAq6UwAA-jaE.jpg   
2050  887343217045368832  https://pbs.twimg.com/ext_tw_video_thumb/88734...   
2051  887473957103951883    https://pbs.twimg.com/media/DFDw2tyUQAAAFke.jpg   
2052  887517139158093824  https://pbs.twimg.com/ext_tw_video_thumb/88751...   
2053  887705289381826560    https://pbs.twimg.com/media/DFHDQBbXgAEqY7t.jpg   
2054  888078434458587136    https://pbs.twimg.com/media/DFMWn56WsAAkA7B.jpg   
2055  888202515573088257    https://pbs.twimg.com/media/DFDw2tyUQAAAFke.jpg   
2056  888554962724278272    https://pbs.twimg.com/media/DFTH_O-UQAACu20.jpg   
2057  888804989199671297    https://pbs.twimg.com/media/DFWra-3VYAA2piG.jpg   
2058  888917238123831296    https://pbs.twimg.com/media/DFYRgsOUQAARGhO.jpg   
2059  889278841981685760  https://pbs.twimg.com/ext_tw_video_thumb/88927...   
2060  889531135344209921    https://pbs.twimg.com/media/DFg_2PVW0AEHN3p.jpg   
2061  889638837579907072    https://pbs.twimg.com/media/DFihzFfXsAYGDPR.jpg   
2062  889665388333682689    https://pbs.twimg.com/media/DFi579UWsAAatzw.jpg   
2063  889880896479866881    https://pbs.twimg.com/media/DFl99B1WsA

In [33]:
# Programmatic Assessment - Information
img_pre.info()

<class 'pandas.core.frame.DataFrame'>
RangeIndex: 2075 entries, 0 to 2074
Data columns (total 12 columns):
tweet_id    2075 non-null int64
jpg_url     2075 non-null object
img_num     2075 non-null int64
p1          2075 non-null object
p1_conf     2075 non-null float64
p1_dog      2075 non-null bool
p2          2075 non-null object
p2_conf     2075 non-null float64
p2_dog      2075 non-null bool
p3          2075 non-null object
p3_conf     2075 non-null float64
p3_dog      2075 non-null bool
dtypes: bool(3), float64(3), int64(2), object(4)
memory usage: 152.1+ KB


In [34]:
# checking for duplicates
img_pre[img_pre['jpg_url'].duplicated(keep = False)].sort_values(by = 'jpg_url')[['tweet_id', 'jpg_url']]

tweet_id                                            jpg_url
480   675354435921575936  https://pbs.twimg.com/ext_tw_video_thumb/67535...
1297  752309394570878976  https://pbs.twimg.com/ext_tw_video_thumb/67535...
1864  842892208864923648  https://pbs.twimg.com/ext_tw_video_thumb/80710...
1641  807106840509214720  https://pbs.twimg.com/ext_tw_video_thumb/80710...
1703  817181837579653120  https://pbs.twimg.com/ext_tw_video_thumb/81596...
1691  815966073409433600  https://pbs.twimg.com/ext_tw_video_thumb/81596...
1705  817423860136083457  https://pbs.twimg.com/ext_tw_video_thumb/81742...
1858  841833993020538882  https://pbs.twimg.com/ext_tw_video_thumb/81742...
1715  819004803107983360    https://pbs.twimg.com/media/C12whDoVEAALRxa.jpg
1718  819015337530290176    https://pbs.twimg.com/media/C12whDoVEAALRxa.jpg
1716  819006400881917954    https://pbs.twimg.com/media/C12x-JTVIAAzdfl.jpg
1717  819015331746349057    https://pbs.twimg.com/media/C12x-JTVIAAzdfl.jpg
1746  823269594223824897    https://pbs.twimg.com/media/C2kzTGxWEAEOpPL.jpg
1738  822244816520155136    https://pbs.twimg.com/media/C2kzTGxWEAEOpPL.jpg
1740  822489057087389700    https://pbs.twimg.com/media/C2oRbOuWEAAbVSl.jpg
1742  822647212903690241    https://pbs.twimg.com/media/C2oRbOuWEAAbVSl.jpg
1767  826958653328592898    https://pbs.twimg.com/media/C3nygbBWQAAjwcW.jpg
1789  829878982036299777    https://pbs.twimg.com/media/C3nygbBWQAAjwcW.jpg
1785  829374341691346946    https://pbs.twimg.com/media/C4KHj-nWQAA3poV.jpg
1903  851953902622658560    https://pbs.twimg.com/media/C4KHj-nWQAA3poV.jpg
2041  885311592912609280    https://pbs.twimg.com/media/C4bTH6nWMAAX_bJ.jpg
1791  830583320585068544    https://pbs.twimg.com/media/C4bTH6nWMAAX_bJ.jpg
224   670319130621435904    https://pbs.twimg.com/media/CU1zsMSUAAAS0qW.jpg
1345  759159934323924993    https://pbs.twimg.com/media/CU1zsMSUAAAS0qW.jpg
1590  798644042770751489    https://pbs.twimg.com/media/CU3mITUWIAAfyQS.jpg
241   670444955656130560    https://pbs.twimg.com/media/CU3mITUWIAAfyQS.jpg
1589  798628517273620480    https://pbs.twimg.com/media/CUN4Or5UAAAa5K4.jpg
85    667509364010450944    https://pbs.twimg.com/media/CUN4Or5UAAAa5K4.jpg
327   671896809300709376    https://pbs.twimg.com/media/CVMOlMiWwAA4Yxl.jpg
1591  798665375516884993    https://pbs.twimg.com/media/CVMOlMiWwAA4Yxl.jpg
...                  ...                                                ...
1571  794983741416415232    https://pbs.twimg.com/media/CvT6IV6WEAQhhV5.jpg
1530  789530877013393408    https://pbs.twimg.com/media/CvT6IV6WEAQhhV5.jpg
1538  790723298204217344    https://pbs.twimg.com/media/CvaYgDOWgAEfjls.jpg
1533  789986466051088384    https://pbs.twimg.com/media/CvaYgDOWgAEfjls.jpg
1683  813944609378369540    https://pbs.twimg.com/media/Cveg1-NXgAASaaT.jpg
1534  790277117346975746    https://pbs.twimg.com/media/Cveg1-NXgAASaaT.jpg
1539  790946055508652032    https://pbs.twimg.com/media/CvoBPWRWgAA4het.jpg
1699  816829038950027264    https://pbs.twimg.com/media/CvoBPWRWgAA4het.jpg
1544  791672322847637504    https://pbs.twimg.com/media/CvyVxQRWEAAdSZS.jpg
1564  793614319594401792    https://pbs.twimg.com/media/CvyVxQRWEAAdSZS.jpg
1804  832215726631055365    https://pbs.twimg.com/media/CwJR1okWIAA6XMp.jpg
1561  793286476301799424    https://pbs.twimg.com/media/CwJR1okWIAA6XMp.jpg
1566  793962221541933056    https://pbs.twimg.com/media/CwS4aqZXUAAe3IO.jpg
1652  809808892968534016    https://pbs.twimg.com/media/CwS4aqZXUAAe3IO.jpg
1572  795076730285391872    https://pbs.twimg.com/media/CwiuEJmW8AAZnit.jpg
1755  824796380199809024    https://pbs.twimg.com/media/CwiuEJmW8AAZnit.jpg
1579  796177847564038144    https://pbs.twimg.com/media/Cwx99rpW8AMk_Ie.jpg
1578  796149749086875649    https://pbs.twimg.com/media/Cwx99rpW8AMk_Ie.jpg
1610  801167903437357056    https://pbs.twimg.com/media/Cx5R8wPVEAALa9r.jpg
1645  808134635716833280    https://pbs.twimg.com/media/Cx5R8wPVEAALa9r.jpg
1603  800141422401830912    https://pbs.twimg.com/media/CxqsX-8X

2. **`image-predictions.tsv`** table

***1 Content Issues:***

**1.1 Visual Assessment:**
- We have few dog breeds that are represented in lowercase.

**1.2 Programmatic Assessment:**
- We have about 281 images on a whole, that are missing with respect to our `twitter-archive-enhanced.csv` file - **Data Quality Dimension - `Completeness`**.
- We can see that, we have about `66` duplicates **OR** a pair of tweets are pointing to same *`jpg_url`* - **Data Quality Dimension - `Accuracy`**.

***2 Structural Issues:***

**2.1 Visual Assessment:**
- None. 

**2.2 Programmatic Assessment:**
- None.

In [35]:
# assessing tweet_master dataset
tweet_master

favorites  retweets            tweet_id
0         38182      8360  892420643555336193
1         32741      6175  892177421306343426
2         24654      4085  891815181378084864
3         41510      8495  891689557279858688
4         39692      9197  891327558926688256
5         19927      3057  891087950875897856
6         11650      2027  890971913173991426
7         64406     18544  890729181411237888
8         27376      4200  890609185150312448
9         31419      7259  890240255349198849
10        30206      7212  890006608113172480
11        27378      4892  889880896479866881
12        47385      9881  889665388333682689
13        26749      4470  889638837579907072
14        14882      2207  889531135344209921
15        24870      5281  889278841981685760
16        28663      4416  888917238123831296
17        25175      4223  888804989199671297
18        19534      3490  888554962724278272
19        21430      3438  888078434458587136
20        29723      5298  887705289381826560
21        45578     11471  887517139158093824
22        68024     17866  887473957103951883
23        33175     10234  887343217045368832
24        30108      5867  887101392804085760
25        34631      7641  886983233522544640
26        11874      3226  886736880519319552
27        22102      4392  886680336477933568
28        20875      3144  886366144734445568
29          116         4  886267009285017600
...         ...       ...                 ...
2312        166        84  666396247373291520
2313        186        92  666373753744588802
2314        770       564  666362758909284353
2315        217        71  666353288456101888
2316        288       133  666345417576210432
2317        196        90  666337882303524864
2318        494       352  666293911632134144
2319        146        64  666287406224695296
2320        174        76  666273097616637952
2321        102        35  666268910803644416
2322      14148      6474  666104133288665088
2323         80        12  666102155909144576
2324        153        67  666099513787052032
2325        163        72  666094000022159362
2326        116        44  666082916733198337
2327        318       162  666073100786774016
2328        145        59  666071193221509120
2329        470       215  666063827256086533
2330        110        57  666058600524156928
2331        292       140  666057090499244032
2332        430       242  666055525042405380
2333       1207       841  666051853826850816
2334        132        59  666050758794694657
2335        108        41  666049248165822465
2336        295       139  666044226329800704
2337        124        43  666033412701032449
2338        128        47  666029285002620928
2339       2539       507  666020888022790149
2340       6076      1741  758740312047005698
2341       2302       857  676957860086095872

[2342 rows x 3 columns]

In [36]:
tweet_master.info()

<class 'pandas.core.frame.DataFrame'>
RangeIndex: 2342 entries, 0 to 2341
Data columns (total 3 columns):
favorites    2342 non-null int64
retweets     2342 non-null int64
tweet_id     2342 non-null int64
dtypes: int64(3)
memory usage: 55.0 KB


3. **`tweet_master.txt`** table

***1 Content Issues:***

**1.1 Visual Assessment:**
- None.

**1.2 Programmatic Assessment:**
- We have about 14 missing records - **Data Quality Dimension - `Completeness`**.

***2 Structural Issues:***

**2.1 Visual Assessment:**
- None.

**2.2 Programmatic Assessment:**
- None.

## Summary - Assessing

- Completed the second step.
- The following are the insights:
    - from `twitter-archive-enhanced.csv` datset, the rating_numerator and denominator need to be fixed.
    - the dataset also represents row values as columns, which needs to be fixed.
    - the dataset also has structural issues such as wrong datatype assigned to a column.
    - from `images-preductions.tsv` dataset, there is consistency issue with naming dog breeds.
    - the dataset isn't complete when compared to `twitter-archive-enhanced.csv`, we have about 281 missing tweets.
    - Also we have `jpg_urls'` that are pointing to a pair of tweets.
    - `tweet_master.txt` dataset has about 14 missing records.
    - the dataset alone hold the information about retweets and favourites - bad form of schema normalization.

## Cleaning

#### Define
- Important!, before we get to cleaning, let's drop rows from image-predictions, that are false in dog_1,_2 and _3, as they are not related to our dataset.

#### Code

In [37]:
# only select those rows that are either true or false and not all false
img_pre = img_pre[~((img_pre.p1_dog == False) & (img_pre.p2_dog == False) & (img_pre.p3_dog == False))]

####  Test

In [38]:
# asserting the lenght to be 0
assert len(img_pre[(img_pre.p1_dog == False) & (img_pre.p2_dog == False) & (img_pre.p3_dog == False)]) == 0, "Check"

In [39]:
# the master dataset
master_set = archive.merge(img_pre, how = 'left', on = ['tweet_id'])
master_set = master_set.merge(tweet_master, how = 'left', on = ['tweet_id'])
files_list.append('dataset/master_set_raw.csv')
master_set.info()

<class 'pandas.core.frame.DataFrame'>
Int64Index: 2356 entries, 0 to 2355
Data columns (total 30 columns):
tweet_id                      2356 non-null int64
in_reply_to_status_id         78 non-null float64
in_reply_to_user_id           78 non-null float64
timestamp                     2356 non-null object
source                        2356 non-null object
text                          2356 non-null object
retweeted_status_id           181 non-null float64
retweeted_status_user_id      181 non-null float64
retweeted_status_timestamp    181 non-null object
expanded_urls                 2297 non-null object
rating_numerator              2356 non-null int64
rating_denominator            2356 non-null int64
name                          2356 non-null object
doggo                         2356 non-null object
floofer                       2356 non-null object
pupper                        2356 non-null object
puppo                         2356 non-null object
jpg_url                       17

In [40]:
# saving the file to local disk.
master_set.to_csv(files_list[-1], index = False)

In [41]:
# creating a copy of the master set
master_set = pd.read_csv(files_list[-1], encoding = 'utf-8')
master_copy = master_set.copy()

### Issues to Clean.

#### 1. Basic cleaning.

#### Define
- Assign proper class for the above 14 tweets before melting.
- Delete *retweets* with *any duplicates* and get rid of *tweets with **no** images*.
- Once done, drop the following columns:
    1. `retweeted_status_id`
    2. `retweeted_status_user_id`
    3. `retweeted_status_timestamp`
    4. `in_reply_to_status_id`
    5. `in_reply_to_user_id`
    
#### Code

In [42]:
# setting column width to -1
pd.set_option('display.max_colwidth', -1)
cond_1 = (master_copy['doggo'] == 'doggo') & (master_copy['floofer'] == 'floofer')
cond_2 = (master_copy['doggo'] == 'doggo') & (master_copy['pupper'] == 'pupper')
cond_3 = (master_copy['doggo'] == 'doggo') & (master_copy['puppo'] == 'puppo')
print(master_copy[cond_1 | cond_2 | cond_3][['tweet_id', 'text']])

                tweet_id  \
191   855851453814013952   
200   854010172552949760   
460   817777686764523521   
531   808106460588765185   
565   802265048156610565   
575   801115127852503040   
705   785639753186217984   
733   781308096455073793   
778   775898661951791106   
822   770093767776997377   
889   759793422261743616   
956   751583847268179968   
1063  741067306818797568   
1113  733109485275860992   

                                                                                                                                                                      text  
191   Here's a puppo participating in the #ScienceMarch. Cleverly disguising her own doggo agenda. 13/10 would keep the planet habitable for https://t.co/cMhq16isel        
200   At first I thought this was a shy doggo, but it's actually a Rare Canadian Floofer Owl. Amateurs would confuse the two. 11/10 only send dogs https://t.co/TXdT3tmuYk  
460   This is Dido. She's playing the lead role in "Pupper S

***Assign the following:***
1. 855851453814013952: puppo
2. 854010172552949760: floofer
3. 817777686764523521: pupper
4. 808106460588765185: pupper
5. 802265048156610565: pupper
6. 801115127852503040: pupper
7. 785639753186217984: pupper
8. 781308096455073793: pupper
9. 775898661951791106: pupper
10. 770093767776997377: pupper
11. 759793422261743616: pupper
12. 751583847268179968: doggo
13. 741067306818797568: doggo
14. 733109485275860992: doggo

**<span style="color: green">I like puppies, so for most of the entries it is pupper!</span>**

In [43]:
# assigning values.
master_copy.loc[master_copy['tweet_id'] == 855851453814013952, ['doggo', 'floofer', 'pupper']] = 'None'
master_copy.loc[master_copy['tweet_id'] == 854010172552949760, ['doggo', 'pupper', 'puppo']] = 'None'
master_copy.loc[master_copy['tweet_id'] == 817777686764523521, ['doggo', 'floofer', 'puppo']] = 'None'
master_copy.loc[master_copy['tweet_id'] == 808106460588765185, ['doggo', 'floofer', 'puppo']] = 'None'
master_copy.loc[master_copy['tweet_id'] == 802265048156610565, ['doggo', 'floofer', 'puppo']] = 'None'
master_copy.loc[master_copy['tweet_id'] == 801115127852503040, ['doggo', 'floofer', 'puppo']] = 'None'
master_copy.loc[master_copy['tweet_id'] == 785639753186217984, ['doggo', 'floofer', 'puppo']] = 'None'
master_copy.loc[master_copy['tweet_id'] == 781308096455073793, ['doggo', 'floofer', 'puppo']] = 'None'
master_copy.loc[master_copy['tweet_id'] == 775898661951791106, ['doggo', 'floofer', 'puppo']] = 'None'
master_copy.loc[master_copy['tweet_id'] == 770093767776997377, ['doggo', 'floofer', 'puppo']] = 'None'
master_copy.loc[master_copy['tweet_id'] == 759793422261743616, ['doggo', 'floofer', 'puppo']] = 'None'
master_copy.loc[master_copy['tweet_id'] == 751583847268179968, ['pupper', 'floofer', 'puppo']] = 'None'
master_copy.loc[master_copy['tweet_id'] == 741067306818797568, ['pupper', 'floofer', 'puppo']] = 'None'
master_copy.loc[master_copy['tweet_id'] == 733109485275860992, ['pupper', 'floofer', 'puppo']] = 'None'

####  Test - 1

In [44]:
# all values have been properly assigned
pd.set_option('display.max_colwidth', 50)
master_copy[cond_1 | cond_2 | cond_3][['tweet_id', 'doggo', 'floofer', 'pupper', 'puppo']]

tweet_id  doggo  floofer  pupper  puppo
191   855851453814013952   None     None    None  puppo
200   854010172552949760   None  floofer    None   None
460   817777686764523521   None     None  pupper   None
531   808106460588765185   None     None  pupper   None
565   802265048156610565   None     None  pupper   None
575   801115127852503040   None     None  pupper   None
705   785639753186217984   None     None  pupper   None
733   781308096455073793   None     None  pupper   None
778   775898661951791106   None     None  pupper   None
822   770093767776997377   None     None  pupper   None
889   759793422261743616   None     None  pupper   None
956   751583847268179968  doggo     None    None   None
1063  741067306818797568  doggo     None    None   None
1113  733109485275860992  doggo     None    None   None

In [45]:
# selecting those, tweets that have no retweets
master_copy = master_copy[pd.isnull(master_copy['retweeted_status_id'])]

# deleting duplicates if any
master_copy = master_copy.drop_duplicates()

# deleting those tweets with no images.
master_copy = master_copy.dropna(subset = ['jpg_url'])

# reseting index
master_copy.reset_index(drop=True, inplace=True)

# droping columns
master_copy = master_copy.drop(labels = ['retweeted_status_id',
                                         'retweeted_status_user_id',
                                         'retweeted_status_timestamp',
                                         'in_reply_to_status_id', 
                                         'in_reply_to_user_id'],
                               axis = 1)

#### Test - 2

In [46]:
# after droping the columns, we should have about 25 dimensions/columns
master_copy.shape

(1685, 25)

#### 2. Condense wide-format to long-format

#### Define
- Condense `doggo`, `floofer`, `pupper`, `puppo` as `dog_class`.

#### Code

In [47]:
# to make sure that we have 
doggo = master_copy.doggo.value_counts()['doggo']
floofer = master_copy.floofer.value_counts()['floofer']
pupper = master_copy.pupper.value_counts()['pupper']
puppo = master_copy.puppo.value_counts()['puppo']

# printing count of each class
print("Count of Doggo: {}\nCount of Floofer: {}\nCount of Pupper: {}\nCount of Puppo: {}".format(doggo,
                                                                                                 floofer,
                                                                                                 pupper,
                                                                                                 puppo))

Count of Doggo: 57
Count of Floofer: 8
Count of Pupper: 173
Count of Puppo: 22


In [48]:
# selecting the columns that are to be melted
columns_to_melt = ['doggo', 'floofer', 'pupper', 'puppo']
columns_to_stay = [x for x in master_copy.columns.tolist() if x not in columns_to_melt]

# melting the the columns into values
master_copy = pd.melt(master_copy, id_vars = columns_to_stay, value_vars = columns_to_melt, 
                         var_name = 'stages', value_name = 'dog_class')

# Delete column 'stages'
master_copy = master_copy.drop('stages', 1)

# dropping duplicates
master_copy = master_copy.sort_values('dog_class').drop_duplicates('tweet_id', keep = 'last')
master_copy.reset_index(drop=True, inplace=True)

#### Test

In [49]:
# let's assert
assert doggo == master_copy.dog_class.value_counts()['doggo'], "Some entries are missing"
assert floofer == master_copy.dog_class.value_counts()['floofer'], "Some entries are missing"
assert pupper == master_copy.dog_class.value_counts()['pupper'], "Some entries are missing"
assert puppo == master_copy.dog_class.value_counts()['puppo'], "Some entries are missing"

#### 3. Fix all inaccurate data.

#### Define
- fix names of dogs.
- fix ratings.
- check source column.

#### Code

In [50]:
# Checking source column
master_copy.source.nunique()

3

**<span style="color: red">Okay! only three values, a categorical variable</span>**

In [51]:
import re
# assiging unique values to source.
master_copy['source'] = master_copy['source'].apply(lambda x: re.findall(r'>(.*)<', x)[0])

####  Test - 1

In [52]:
# taking a look at sample of 5 rows
master_copy.sample(5)[['tweet_id', 'source', 'text']]

tweet_id              source  \
28    667062181243039745  Twitter for iPhone   
1348  852912242202992640  Twitter for iPhone   
1420  761004547850530816  Twitter for iPhone   
1322  753026973505581056  Twitter for iPhone   
431   700890391244103680  Twitter for iPhone   

                                                   text  
28    This is Keet. He is a Floridian Amukamara. Abs...  
1348  Meet Benny. He likes being adorable and making...  
1420  This is Bo and Ty. Bo eats paper and Ty felt l...  
1322  Say hello to Tayzie. She's a Barbadian Bugaboo...  
431   This is Chet. He's dapper af. His owners want ...

In [53]:
# fixing names
non_names = master_copy.name.str.islower()
non_names = list(set(master_copy[non_names]['name'].tolist()))
flag = master_copy.name.str.len() == 1 & master_copy.name.str.isupper()
non_names.append(master_copy[flag][['tweet_id', 'name']]['name'].tolist()[0])

In [54]:
# replacing all garbage names with none, once done, we'll use the text field to extract names
for name in master_copy.name:
    if name in non_names:
        master_copy.loc[master_copy['name'] == name, ['name']] = 'None'

In [55]:
# checking if there are any non_names after the operation
assert len(master_copy[(master_copy.name.str.islower()) & (flag)]) == 0, "Check code"

***The following are patterns observed in `text` field, we shall use the :***
- This is [name] ..
- Meet [name] ..
- Say hello to [name] ..
- .. named [name] ..
- .. name is [name] ..

We will treat those cases to get the names from the text of the tweet

In [56]:
# extracting names using regular expression.
dog_names = []

# assigning patterns
pattern_1 = r'(T|t)his\sis\s([^.|,]*)'
pattern_2 = r'Meet\s([^.|,]*)'
pattern_3 = r'Say\shello\sto\s([^.|,]*)'
pattern_4 = r'name\sis\s([^.|,]*)'

# looping through text and extracting names
for text in master_copy['text']:
    # Start with 'This is '
    if re.search(pattern_1, text):
        # if our match has alternate name
        if "(" in re.search(pattern_1, text).group(2):
            dog_names.append(re.search(pattern_1, text).group(2).split()[0])
        # if our match has AKA in it
        elif "AKA" in re.search(pattern_1, text).group(2):
            dog_names.append(re.search(pattern_1, text).group(2).split()[0])
        # if our name has two dogs
        elif '&amp;' in re.search(pattern_1, text).group(2):
            temp = re.search(pattern_1, text).group(2).split()
            if len(temp) == 1:
                dog_names.append(temp[0])
            elif len(temp) == 3:
                dog_names.append(temp[0]+"|"+temp[-1])
            else:
                dog_names.append(temp[0]+"|"+temp[-2])
        elif 'named' in re.search(pattern_1, text).group(2):
            temp = re.search(pattern_1, text).group(2).split()
            dog_names.append(temp[-1])
        # just appending the name
        else:
            dog_names.append(re.search(pattern_1, text).group(2))
    
    # Start with 'Meet '
    elif re.search(pattern_2, text):
        # if our name has two dogs
        if '&amp;' in re.search(pattern_2, text).group(1):
            temp = re.search(pattern_2, text).group(1).split()
            if len(temp) == 1:
                dog_names.append(temp[0])
            elif len(temp) == 3:
                dog_names.append(temp[0]+"|"+temp[-1])
            else:
                dog_names.append(temp[0]+"|"+temp[-2])
        # if our name has alternatives
        elif '(' in re.search(pattern_2, text).group(1):
            dog_names.append(re.search(pattern_2, text).group(1).split()[0])
        # just appending the name
        else:
            dog_names.append(re.search(pattern_2, text).group(1))
    
    # Start with 'Say hello to '
    elif re.search(pattern_3, text):
        # if our match has alternate name
        if '(' in re.search(pattern_3, text).group(1):
            dog_names.append(re.search(pattern_3, text).group(1).split()[0])
        # if our name has two dogs
        elif '&amp;' in re.search(pattern_3, text).group(1):
            temp = re.search(pattern_3, text).group(1).split()
            if len(temp) == 1:
                dog_names.append(temp[0])
            elif len(temp) == 3:
                dog_names.append(temp[0]+"|"+temp[-1])
            else:
                dog_names.append(temp[0]+"|"+temp[-2])
        else:
            dog_names.append(re.search(pattern_3, text).group(1))    
    
    # contains 'name is'
    elif re.search(pattern_4, text):
        if len(re.search(pattern_4, text).group(1).split()) == 1:
            dog_names.append(re.search(pattern_4, text).group(1))
        else:
            temp = re.search(pattern_4, text).group(1).split()
            dog_names.append(temp[0])
        
    # No name specified or other style
    else:
        dog_names.append('None')

# adding this new set of names to our master_copy
master_copy['dog_names'] = dog_names

In [57]:
# new non names.
non_names = []
pattern_4 = r'^[a-z].*'
for name in master_copy['dog_names']:
    if re.search(pattern_4, name):
        master_copy.loc[master_copy['dog_names'] == name, ['dog_names']] = 'None'
        non_names.append(name)

In [58]:
# dog_names with and to be replaced with |
for name in master_copy['dog_names']:
    master_copy['dog_names'] = master_copy['dog_names'].str.replace(pat = r'\sand\s', repl = "|", regex = True)

In [59]:
# we need to replace two cells, with names 'Sadie|&amp;', 'Phillippe ...',
pd.set_option('display.max_colwidth', -1)
master_copy[(master_copy['dog_names'] == "Philippe from Soviet Russia") |
            (master_copy['dog_names'] == "Sadie|&amp;")][['tweet_id', 'text', 'dog_names']]

tweet_id  \
23   667495797102141441   
160  693280720173801472   

                                                                                                                                                text  \
23   This is Philippe from Soviet Russia. Commanding leader. Misplaced other boot. Hung flag himself. 9/10 charismatic af https://t.co/5NhPV8E45i      
160  This is Sadie and her 2 pups Shebang &amp; Ruffalo. Sadie says single parenting is challenging but rewarding. All 10/10 https://t.co/UzbhwXcLne   

                       dog_names  
23   Philippe from Soviet Russia  
160  Sadie|&amp;

In [60]:
# setting them to correct ones
# the following tweet_id we different from our regexs
master_copy.loc[master_copy['dog_names'] == "Philippe from Soviet Russia", ['dog_names']] = 'Phillippe'
master_copy.loc[master_copy['dog_names'] == "Sadie|&amp;", ['dog_names']] = 'Sadie|Shebang|Ruffalo'
master_copy.loc[master_copy['tweet_id'] == 667509364010450944, ['dog_names']] = 'Tickles'
master_copy.loc[master_copy['tweet_id'] == 667546741521195010, ['dog_names']] = 'George'
master_copy.loc[master_copy['tweet_id'] == 667073648344346624, ['dog_names']] = 'Dave'
master_copy.loc[master_copy['tweet_id'] == 667177989038297088, ['dog_names']] = 'Daryl'
master_copy.loc[master_copy['tweet_id'] == 666835007768551424, ['dog_names']] = 'Cupit|Prencer'
master_copy.loc[master_copy['tweet_id'] == 668221241640230912, ['dog_names']] = 'Bo|Smittens'
master_copy.loc[master_copy['tweet_id'] == 668268907921326080, ['dog_names']] = 'Guss'
master_copy.loc[master_copy['tweet_id'] == 666058600524156928, ['dog_names']] = 'Paul Rand'
master_copy.loc[master_copy['tweet_id'] == 692142790915014657, ['dog_names']] = 'Teddy'
master_copy.loc[master_copy['tweet_id'] == 684097758874210310, ['dog_names']] = 'Lupe'
master_copy.loc[master_copy['tweet_id'] == 709198395643068416, ['dog_names']] = 'Cletus|Jerome|Alejandro|Burp|Titson'
master_copy.loc[master_copy['tweet_id'] == 671743150407421952, ['dog_names']] = 'Jacob'
master_copy.loc[master_copy['tweet_id'] == 669037058363662336, ['dog_names']] = 'Pancho|Peaches'
master_copy.loc[master_copy['tweet_id'] == 669363888236994561, ['dog_names']] = 'Zeus'
master_copy.loc[master_copy['tweet_id'] == 813217897535406080, ['dog_names']] = 'Atlas'
master_copy.loc[master_copy['tweet_id'] == 856526610513747968, ['dog_names']] = 'Charlie'
master_copy.loc[master_copy['tweet_id'] == 861288531465048066, ['dog_names']] = 'Boomer'
master_copy.loc[master_copy['tweet_id'] == 863079547188785154, ['dog_names']] = 'Pipsy'
master_copy.loc[master_copy['tweet_id'] == 844979544864018432, ['dog_names']] = 'Toby'
master_copy.loc[master_copy['tweet_id'] == 836001077879255040, ['dog_names']] = 'Atlas'
master_copy.loc[master_copy['tweet_id'] == 758041019896193024, ['dog_names']] = 'Teagan'
master_copy.loc[master_copy['tweet_id'] == 765395769549590528, ['dog_names']] = 'Zoey'
master_copy.loc[master_copy['tweet_id'] == 778408200802557953, ['dog_names']] = 'Loki'
master_copy.loc[master_copy['tweet_id'] == 770069151037685760, ['dog_names']] = 'Carbon'

In [61]:
# dropping column name
master_copy = master_copy.drop(['name'], axis = 1)

# printing columns in master_copy
master_copy.columns.tolist()

['tweet_id',
 'timestamp',
 'source',
 'text',
 'expanded_urls',
 'rating_numerator',
 'rating_denominator',
 'jpg_url',
 'img_num',
 'p1',
 'p1_conf',
 'p1_dog',
 'p2',
 'p2_conf',
 'p2_dog',
 'p3',
 'p3_conf',
 'p3_dog',
 'favorites',
 'retweets',
 'dog_class',
 'dog_names']

#### Test - 2

In [62]:
# selecting a dog_names that are none.
master_copy[master_copy['dog_names'] == 'None'].sample(5)[['tweet_id', 'text', 'dog_names']]

tweet_id  \
776   674646392044941312   
655   670778058496974848   
494   728760639972315136   
1289  740711788199743490   
793   674790488185167872   

                                                                                                                                               text  \
776   Two gorgeous dogs here. Little waddling dog is a rebel. Refuses to look at camera. Must be a preteen. 5/10 &amp; 8/10 https://t.co/YPfw7oahbD   
655   "To bone or not to bone?"\r\r\n10/10 https://t.co/4g5kFdxp6g                                                                                    
494   When you're way too slow for the "down low" portion of a high five. 13/10 https://t.co/Cofwoy7Vpq                                               
1289  Here we are witnessing the touchdown of a pupnado. It's not funny it's actually very deadly. 9/10 might still pet https://t.co/CmLoKMbOHv       
793   ER... MER... GERD 13/10 https://t.co/L1puJISV1a                                                                                                 

     dog_names  
776   None      
655   None      
494   None      
1289  None      
793   None

####  Define
- Let's get cleaning the ratings.

#### Code

In [63]:
# as we are aware that there are two ratings in the text columns, lets use the
# regex to extract and replace the wrong ones.
ratings = master_copy['text'].apply(lambda x: re.findall(r'(\d+(\.\d+)|(\d+))\/(\d+0)', x))

In [64]:
# let's scale number from 0 to 15
def scale_rate(number, mini = 0, maxi = 15):
    return (number - mini)/(maxi - mini)

In [65]:
import math
# our temp variables
rate_num = []
rate_denom = []

# let's loop over and assign values properly
for rate in ratings:
    # if our regex didn't return any value.
    if len(rate) == 0:
        rate_num.append(0)
        rate_denom.append(0)
    
    # if regex's leght equals to one
    elif len(rate) == 1:
        
        temp = float(rate[0][0])
        
        # if we the value falls in between  [30,100]
        if 30 < temp < 100:
            temp_2 = int(rate[0][0]) / int(rate[0][-1][0])
            rate_num.append(math.ceil(temp_2))
            rate_denom.append(10)
        
        # else if our number falls between [100, 200]
        elif 100 < temp < 200:
            temp_2 = int(rate[0][0]) / int(rate[0][-1][:2])
            rate_num.append(math.ceil(temp_2))
            rate_denom.append(10)
        
        # else just ceiling the number
        else:
            rate_num.append(math.ceil(temp))
            rate_denom.append(10)
    
    # if our regex returned two ratings
    elif len(rate) == 2:
        
        # restricting our ratings to a max of 15
        if int(rate[0][0]) + int(rate[1][0]) > 15:
            temp = (int(rate[0][0]) + int(rate[1][0]))/2
            rate_num.append(math.ceil(temp))
            rate_denom.append(10)
        # if it is < 15
        else:
            rate_num.append(int(rate[0][0]) + int(rate[1][0]))
            rate_denom.append(10)
    
    # all others lenghts
    else:
        temp_sum = 0
        for i in range(len(rate)):
            temp_sum += int(rate[i][0])
        
        scaled = scale_rate(temp_sum)
        
        rate_num.append(math.ceil(scaled))
        rate_denom.append(10)

# assigning the values to rating_numerator and denominator
master_copy['rating_numerator'] = rate_num
master_copy['rating_denominator'] = rate_denom

#### Test

In [66]:
# lenght of the newly obtained values equal to the number of datapoint in our dataset
# i.e. 1686
print(len(master_copy['rating_numerator']), len(master_copy['rating_denominator']))

1685 1685


#### 4. Fix the structure of the table

#### Define
- Assign each column with appropriate type

#### Code

In [67]:
# info about master_copy
master_copy.info()

<class 'pandas.core.frame.DataFrame'>
RangeIndex: 1685 entries, 0 to 1684
Data columns (total 22 columns):
tweet_id              1685 non-null int64
timestamp             1685 non-null object
source                1685 non-null object
text                  1685 non-null object
expanded_urls         1685 non-null object
rating_numerator      1685 non-null int64
rating_denominator    1685 non-null int64
jpg_url               1685 non-null object
img_num               1685 non-null float64
p1                    1685 non-null object
p1_conf               1685 non-null float64
p1_dog                1685 non-null object
p2                    1685 non-null object
p2_conf               1685 non-null float64
p2_dog                1685 non-null object
p3                    1685 non-null object
p3_conf               1685 non-null float64
p3_dog                1685 non-null object
favorites             1685 non-null float64
retweets              1685 non-null float64
dog_class             1685 non

In [68]:
# importing numpy
import numpy as np

# assign each column with appropriate type
master_copy['tweet_id'] = master_copy['tweet_id'].astype(object)
master_copy['timestamp'] = pd.to_datetime(master_copy.timestamp)
master_copy['source'] = master_copy['source'].astype('category')
master_copy['favorites'] = master_copy['favorites'].astype(np.int64)
master_copy['retweets'] = master_copy['retweets'].astype(np.int64)
master_copy['dog_class'] = master_copy['dog_class'].astype('category')
master_copy['rating_numerator'] = master_copy['rating_numerator'].astype(np.int64)
master_copy['rating_denominator'] = master_copy['rating_denominator'].astype(np.int64)

#### Test

In [69]:
# printing information
master_copy.info()

<class 'pandas.core.frame.DataFrame'>
RangeIndex: 1685 entries, 0 to 1684
Data columns (total 22 columns):
tweet_id              1685 non-null object
timestamp             1685 non-null datetime64[ns]
source                1685 non-null category
text                  1685 non-null object
expanded_urls         1685 non-null object
rating_numerator      1685 non-null int64
rating_denominator    1685 non-null int64
jpg_url               1685 non-null object
img_num               1685 non-null float64
p1                    1685 non-null object
p1_conf               1685 non-null float64
p1_dog                1685 non-null object
p2                    1685 non-null object
p2_conf               1685 non-null float64
p2_dog                1685 non-null object
p3                    1685 non-null object
p3_conf               1685 non-null float64
p3_dog                1685 non-null object
favorites             1685 non-null int64
retweets              1685 non-null int64
dog_class             1

#### 5. Getting rid of predictions and add in final touches

#### Define
- Condense the `p_[1|2|3]` to `predicted_dog` and `conf_[1|2|3]` to `accuracy`.
- drop columns `img_num`, `p1`, `p1_conf`, `p1_dog`, `p2`, `p2_conf`, `p2_dog`, `p3`, `p3_conf`, `p3_dog`.
- rename columns that are apt to this dataset.

#### Code

In [70]:
# We will store the fisrt true algorithm
# with it's level of confidence
predicted_dog_breed = []
accuracy = []

# funvtion for getting the levels
def condense_predictions(dataframe):
    '''
    takes in the dataframe and extracts information for predicted dog breed.
    dataframe: input to the fuction
    '''
    if dataframe['p1_dog'] == True:
        predicted_dog_breed.append(dataframe['p1'])
        accuracy.append(dataframe['p1_conf'])
    elif dataframe['p2_dog'] == True:
        predicted_dog_breed.append(dataframe['p2'])
        accuracy.append(dataframe['p2_conf'])
    elif dataframe['p3_dog'] == True:
        predicted_dog_breed.append(dataframe['p3'])
        accuracy.append(dataframe['p3_conf'])
    else:
        predicted_dog_breed.append('NaN')
        accuracy.append(0)

master_copy.apply(condense_predictions, axis=1)
master_copy['dog_breeds'] = predicted_dog_breed
master_copy['accuracy'] = accuracy

In [71]:
# dropping columns
master_copy.drop(['img_num',
                  'p1', 'p1_conf', 'p1_dog',
                  'p2', 'p2_conf', 'p2_dog',
                  'p3', 'p3_conf', 'p3_dog'],
                 axis = 1,
                 inplace = True)

In [72]:
# final tocuh - renaming
master_copy.rename(columns={'source': 'tweet_source',
                            'text': 'tweet',
                            'timestamp': 'tweet_date',
                            'expanded_urls' : 'tweet_urls',
                            'jpg_url': 'image_url',
                            'favorites': 'tweet_favorites',
                            'retweets': 'tweet_retweets'},
                   inplace = True)

####  Test

In [73]:
# printing information
master_copy.info()

<class 'pandas.core.frame.DataFrame'>
RangeIndex: 1685 entries, 0 to 1684
Data columns (total 14 columns):
tweet_id              1685 non-null object
tweet_date            1685 non-null datetime64[ns]
tweet_source          1685 non-null category
tweet                 1685 non-null object
tweet_urls            1685 non-null object
rating_numerator      1685 non-null int64
rating_denominator    1685 non-null int64
image_url             1685 non-null object
tweet_favorites       1685 non-null int64
tweet_retweets        1685 non-null int64
dog_class             1685 non-null category
dog_names             1685 non-null object
dog_breeds            1685 non-null object
accuracy              1685 non-null float64
dtypes: category(2), datetime64[ns](1), float64(1), int64(4), object(6)
memory usage: 161.6+ KB


## Storing

**Done with the process of cleaning, let's store and analyse this dataset.**

In [74]:
# saving the file to local disk
master_copy.to_csv(folder+'/'+'twitter_archive_master.csv', index=False, encoding = 'utf-8')

# listing the dataset folder
os.listdir(folder)

['archive_copy.csv',
 'image-predictions.tsv',
 'master_set_raw.csv',
 'tweet_json.txt',
 'tweet_master.csv',
 'twitter-archive-enhanced.csv',
 'twitter_archive_master.csv']